libs/stokesian_dynamics/sd.hpp

```diff
@@ -0,0 +1,1478 @@
+    mobility matrix. The subscript UF indicates that this tensor describes
+    the relationship between forces and velocities.
+    The mobility matrix depends on the positions of all particles and therefore
+    is only valid for the momentary configuration. It needs to be recalculated
+    each time step.
+
+    To obtain the velocities, the mobility matrix must be applied to the forces
+    that act on the particles, giving the velocities.
+
+    These are the steps that are taken to compute the mobility matrix:
+      -# Starting point is an empty mobility matrix.
+      -# All self mobility terms are added. Its translational part is widely
+         known as Stoke's Law.
+      -# All pair mobility terms are added.
+      -# The mobility matrix is inverted to become a resistance matrix.
+      -# Lubrication corrections are added to the resistance matrix. They are
+         necessary for particles in close vicinity because simplifications
+         have been made to efficiently cover long range interactions.
+      -# In the end, the finished resistance matrix is inverted again. The
+         result is the finished mobility matrix that includes the short-ranged
+         lubrication interactions.
+
+    The thermalisation is achieved by a generalization of the 
+    Einstein-Smoluchowski equation, which relates mobility and diffusion: 
+
+            D = \mu k_B T
+
+    where \mu is the mobility and D is the diffusion coefficient. The mean 
+    square displacement during a time step of length \Delta t, and along one
+    degree of freedom is given by
+
+            \langle x^2 \rangle / \Delta t = 2 D
+
+    That way we can determine the distribution of the random displacement
+    that our system experiences along one of its many degrees of freedom. In
+    our case, \mu is a matrix and we need its square root. The latter is 
+    obtained via cholesky decomposition.
+ */
+template <typename Policy, typename T>
+struct solver {
+    static_assert(policy::is_policy<Policy>::value,
+                  "The execution policy must meet the requirements");
+    static_assert(std::is_arithmetic<T>::value,
+                  "Data type of device_matrix must be arithmetic for "
+                  "arithmetic operations");
+
+    template <typename U>
+    using vector_type = typename Policy::template vector<U>;
+    
+    /** viscosity of the Stokes fluid */
+    T const eta;
+    /** number of particles */
+    std::size_t const n_part;
+    /** number of pairs of particles = n_part*(n_part-1)/2 */
+    std::size_t const n_pair;
+
+    device_matrix<T, Policy> zmuf;
+    device_matrix<T, Policy> zmus;
+    device_matrix<T, Policy> zmes;
+
+    solver(T eta, std::size_t const n_part)
+        : eta{eta}, n_part(n_part), n_pair(n_part * (n_part - 1) / 2),
+          zmuf(n_part * 6, n_part * 6), zmus(n_part * 6, n_part * 5),
+          zmes(n_part * 5, n_part * 5) {}
+
+    std::vector<T> calc_vel(std::vector<T> const &x_host,
```

Comment:
Please split this function into smaller methods along the lines of the comment mentioning the high-level steps of the method (Line 1238 or so)